### PR TITLE
[Test Improver] test: add unit tests for LoggingManager.BuildAsync and ExtensionValidationHelper.ValidateUniqueExtension

### PR DESCRIPTION
**Pull request by Evangelink**
Adds focused unit test coverage in `Microsoft.Testing.Platform.UnitTests` for:

- `LoggingManager.BuildAsync` (provider inclusion/exclusion, initialization behavior, and factory argument propagation)
- `ExtensionValidationHelper.ValidateUniqueExtension` (null guards, duplicate UID detection, and error message assertions)

Also includes follow-up review feedback fixes:
- Renamed the duplicate-UID type-name test to match what it asserts.
- Standardized logger category literals in `LoggingManagerTests`.
- Scoped test-only helper interfaces inside `LoggingManagerTests`.
- Added explicit `InitializeAsync` setup in the disabled-provider test to keep regression failures actionable.

Finally, this PR branch was merged with `main` and merge conflicts were resolved in the two new test files without changing the intended test coverage scope.

### CODE REVIEW

File: test/UnitTests/Microsoft.Testing.Platform.UnitTests/Helpers/ExtensionValidationHelperTests.cs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        TestExtension newExtension = new("uid1");
+
+        InvalidOperationException ex = Assert.ThrowsExactly<InvalidOperationException>(
+            () => existing.ValidateUniqueExtension(newExtension));
+
+        Assert.Contains("uid1", ex.Message);
+    }
+
+    [TestMethod]
+    public void ValidateUniqueExtension_WithDuplicateUid_ErrorMessageContainsBothTypeNames()
```
Comment (Evangelink) [inline test/UnitTests/Microsoft.Testing.Platform.UnitTests/Helpers/ExtensionValidationHelperTests.cs]: **Nitpick (Important)**: The test name `ErrorMessageContainsBothTypeNames` implies the message is checked for two distinct type names, but the assertion on line 78 only verifies `typeof(TestExtension).ToString()` — and both `existing[0]` and `newExtension` are the same type, so only one type name is ever asserted.

Either rename the test to `ErrorMessageContainsTypeName`, or introduce a second extension type (e.g. a `TestExtension2`) so you can truly verify both types appear in the error message.

File: test/UnitTests/Microsoft.Testing.Platform.UnitTests/Logging/LoggingManagerTests.cs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        enabledExtProvider.Setup(p => p.CreateLogger(It.IsAny<string>())).Returns(mockLogger.Object);
+        disabledExtProvider.Setup(p => p.IsEnabledAsync()).ReturnsAsync(false);
+
+        LoggingManager manager = new();
+        manager.AddProvider((_, _) => nonExtProvider.Object);
+        manager.AddProvider((_, _) => enabledExtProvider.Object);
+        manager.AddProvider((_, _) => disabledExtProvider.Object);
+
+        ILoggerFactory factory = await manager.BuildAsync(_mockServiceProvider.Object, LogLevel.Information, _mockMonitor.Object);
+        _ = factory.CreateLogger("cat");
```
Comment (Evangelink) [inline test/UnitTests/Microsoft.Testing.Platform.UnitTests/Logging/LoggingManagerTests.cs]: **Nitpick (Minor)**: Category name `"cat"` is an unusual abbreviation — the other tests in this file use `"category"` consistently. Consider using `"category"` here as well for uniformity.

File: test/UnitTests/Microsoft.Testing.Platform.UnitTests/Logging/LoggingManagerTests.cs [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        => _mockMonitor.Setup(x => x.Lock(It.IsAny<object>())).Returns(new Mock<IDisposable>().Object);
+
+    [TestMethod]
+    public async Task BuildAsync_WithNoProviders_ReturnsFactoryThatCreatesLogger()
+    {
+        LoggingManager manager = new();
+
+        ILoggerFactory factory = await manager.BuildAsync(_mockServiceProvider.Object, LogLevel.Information, _mockMonitor.Object);
+
+        Assert.IsNotNull(factory.CreateLogger("test"));
```
Comment (Evangelink) [inline test/UnitTests/Microsoft.Testing.Platform.UnitTests/Logging/LoggingManagerTests.cs]: **Nitpick (Minor)**: The category string `"test"` is inconsistent with `"category"` used in the other provider tests. Aligning on a single placeholder name across all tests reduces unnecessary variation.

File: test/UnitTests/Microsoft.Testing.Platform.UnitTests/Logging/LoggingManagerTests.cs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    [TestMethod]
+    public void AddProvider_WithNullFactory_ThrowsArgumentNullException()
+    {
+        LoggingManager manager = new();
+
+        Assert.ThrowsExactly<ArgumentNullException>(() => manager.AddProvider(null!));
+    }
+}
+
+internal interface IExtensionLoggerProvider : ILoggerProvider, IExtension;
```
Comment (Evangelink) [inline test/UnitTests/Microsoft.Testing.Platform.UnitTests/Logging/LoggingManagerTests.cs]: **Nitpick (Minor)**: `IExtensionLoggerProvider`, `IExtensionInitializableLoggerProvider`, and `IInitializableLoggerProvider` are declared at file scope but are only used by `LoggingManagerTests`. Nesting them inside the test class would better communicate their test-only scope and prevent accidental misuse in other files within the same test assembly.

**Review by Evangelink — COMMENTED** [review]
### Summary

**Workflow**: PR Nitpick Reviewer 🔍
**Date**: 2026-05-11
**Repository**: microsoft/testfx

### Key Findings

**Important**
- `ValidateUniqueExtension_WithDuplicateUid_ErrorMessageContainsBothTypeNames` asserts only one type name: both `existing[0]` and `newExtension` are the same type (`TestExtension`), so the test name is misleading.

**Minor**
- Inconsistent logger category names across tests: `"test"`, `"category"`, and `"cat"` are used for the same conceptual placeholder.
- Three `internal interface` helpers are declared at file scope; nesting them inside `LoggingManagerTests` would better scope them.

### Positive Highlights

- Clean Arrange/Act/Assert structure throughout both test files.
- Good coverage of `null` parameters, enabled/disabled providers, and initializable variants.
- The modification to `TestExtension` to accept a configurable `uid` via constructor is a well-thought-out change to enable richer test scenarios.

---

*Generated by [PR Nitpick Reviewer](https://github.com/microsoft/testfx/actions/runs/25677103577)*

> 🔍 *Meticulously inspected by [PR Nitpick Reviewer 🔍](https://github.com/microsoft/testfx/actions/runs/25677103577)*

**Review by Evangelink — COMMENTED** [review]
### Summary

**Workflow**: Expert Code Reviewer
**Date**: 2026-05-11
**Repository**: microsoft/testfx

### Review

PR #8125 adds 25 unit tests for two internal components:

- **`LoggingManager.BuildAsync`** (10 tests): Covers provider inclusion/exclusion based on `IExtension.IsEnabledAsync()`, initialization ordering via `IAsyncInitializableExtension.TryInitializeAsync()`, log level and service provider propagation, and null argument validation.
- **`ExtensionValidationHelper.ValidateUniqueExtension`** (15 tests): Covers both overloads (direct and selector-based), null argument guards, empty collections, non-duplicate pass-through, duplicate detection, and error message content.

### Key Findings

No issues found in any of the review categories.

**Verified correct:**
- All async test assertions properly `await` `BuildAsync`, which chains through `IsEnabledAsync()` and `TryInitializeAsync()` correctly.
- `Assert.Contains(substring, value)` parameter order is correct per MSTest's string overload signature (substring first, container second).
- `LoggerFactory.CreateLogger` caches per category — `Times.Once` verifications are accurate.
- `IExtensionLoggerProvider`, `IExtensionInitializableLoggerProvider`, and `IInitializableLoggerProvider` test-internal interface declarations are unique in the namespace and correctly compose `IExtension` + `IAsyncInitializableExtension` to trigger the right branches in `BuildAsync`.
- `TestExtension.cs` refactor (constructor-based `Uid` with default `"Uid"`) is backward-compatible with all existing usages.
- `Mock<IMonitor>` setup in the test constructor returns a valid `IDisposable` for `LoggerFactory`'s `using (_monitor.Lock(...))` pattern.
- Moq 4.18.4 default behavior for `Task`-returning methods returns `Task.CompletedTask` — explicit setups are correct (and redundant ones are harmless).

### Recommendations

None — the tests are well-structured and correctly exercise the production code paths.

---

*Generated by [Expert Code Reviewer](https://github.com/microsoft/testfx/actions/runs/25677104851)*

> 🧠 *Reviewed by [Expert Code Reviewer 🧠](https://github.com/microsoft/testfx/actions/runs/25677104851)*

**Review by Evangelink — COMMENTED** [review]
### Summary

**Workflow**: Test Expert Reviewer 🧪
**Date**: 2026-05-11
**Repository**: microsoft/testfx

### Key Findings

1. **[Assertion]** `ValidateUniqueExtension_WithDuplicateUid_ErrorMessageContainsBothTypeNames` — the test name promises verification of two distinct type names in the error message, but both extensions are `TestExtension` and only one `Assert.Contains` is made. This is misleading: if the production code omitted one type name, the test would still pass.

2. **[Coverage]** `LoggingManagerTests` covers null validation for `AddProvider` but not for `BuildAsync` itself. If `BuildAsync` guards its `serviceProvider` or `monitor` parameters against null, those guards have no test coverage.

### Recommendations

- For finding 1: either rename the test to `...ErrorMessageContainsTypeName` or introduce two different concrete `IExtension` types with two separate `Assert.Contains` calls.
- For finding 2: add `BuildAsync_WithNullServiceProvider_ThrowsArgumentNullException` and `BuildAsync_WithNullMonitor_ThrowsArgumentNullException` if those null guards exist in the production code.

Overall the tests are well-structured, properly isolated (MSTest new-instance-per-test, no shared mutable state), use the correct assertion APIs for this project, and cover the key happy/sad paths for both `LoggingManager.BuildAsync` and `ExtensionValidationHelper.ValidateUniqueExtension`.

---

*Generated by [Test Expert Reviewer](https://github.com/microsoft/testfx/actions/runs/25677104853)*

> 🧪 *Test quality reviewed by [Test Expert Reviewer 🧪](https://github.com/microsoft/testfx/actions/runs/25677104853)*

File: test/UnitTests/Microsoft.Testing.Platform.UnitTests/Logging/LoggingManagerTests.cs [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        ILoggerFactory factory = await manager.BuildAsync(_mockServiceProvider.Object, LogLevel.Information, _mockMonitor.Object);
+        _ = factory.CreateLogger("cat");
+
+        nonExtProvider.Verify(p => p.CreateLogger("cat"), Times.Once);
+        enabledExtProvider.Verify(p => p.CreateLogger("cat"), Times.Once);
+        disabledExtProvider.Verify(p => p.CreateLogger(It.IsAny<string>()), Times.Never);
+    }
+
+    [TestMethod]
+    public void AddProvider_WithNullFactory_ThrowsArgumentNullException()
```
Comment (Evangelink) [inline test/UnitTests/Microsoft.Testing.Platform.UnitTests/Logging/LoggingManagerTests.cs:174]: **[Coverage]** The null-parameter coverage for `LoggingManager` only tests `AddProvider(null!)`. The PR description claims null validation for `BuildAsync`, but no tests cover null `serviceProvider` or null `monitor` passed to `BuildAsync`.

**Impact**: If `BuildAsync` performs null guards on `serviceProvider` or `monitor`, those guards are untested.

**Suggestion**: Add null-parameter tests for `BuildAsync` itself if null checks exist in the production code:
```csharp
[TestMethod]
public async Task BuildAsync_WithNullServiceProvider_ThrowsArgumentNullException()
{
    LoggingManager manager = new();
    await Assert.ThrowsExactlyAsync<ArgumentNullException>(
        () => manager.BuildAsync(null!, LogLevel.Information, _mockMonitor.Object));
}

[TestMethod]
public async Task BuildAsync_WithNullMonitor_ThrowsArgumentNullException()
{
    LoggingManager manager = new();
    await Assert.ThrowsExactlyAsync<ArgumentNullException>(
        () => manager.BuildAsync(_mockServiceProvider.Object, LogLevel.Information, null!));
}
```

File: test/UnitTests/Microsoft.Testing.Platform.UnitTests/Helpers/ExtensionValidationHelperTests.cs [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    {
+        List<IExtension> existing = [new TestExtension("uid1")];
+        TestExtension newExtension = new("uid1");
+
+        InvalidOperationException ex = Assert.ThrowsExactly<InvalidOperationException>(
+            () => existing.ValidateUniqueExtension(newExtension));
+
+        Assert.Contains(typeof(TestExtension).ToString(), ex.Message);
```
Comment (Evangelink) [inline test/UnitTests/Microsoft.Testing.Platform.UnitTests/Helpers/ExtensionValidationHelperTests.cs:78]: **[Assertion]** The test name `ErrorMessageContainsBothTypeNames` implies verifying that two *distinct* type names appear in the error message, but only a single `Assert.Contains` call is made — and both the existing and new extension are the same type (`TestExtension`), so this assertion only proves the message contains that one type name at least once.

**Impact**: False confidence — the test passes even if the error message only mentions one of the two conflicting types. If the production code only includes the existing type's name (not the new one's), this test would not catch the regression.

**Suggestion**: Either rename the test to `ValidateUniqueExtension_WithDuplicateUid_ErrorMessageContainsTypeName`, or introduce a second distinct `IExtension` implementation and add two separate `Assert.Contains` calls to verify both type names are present:
```csharp
// Use two different types to truly test "both type names"
List<IExtension> existing = [new TestExtension("uid1")];
AnotherTestExtension newExtension = new("uid1");

// then assert both type names appear
Assert.Contains(typeof(TestExtension).ToString(), ex.Message);
Assert.Contains(typeof(AnotherTestExtension).ToString(), ex.Message);
```

**Comment by Evangelink** [comment]
@copilot address review comments

File: test/UnitTests/Microsoft.Testing.Platform.UnitTests/Logging/LoggingManagerTests.cs [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        _ = factory.CreateLogger("category");
+
+        mockProvider.Verify(p => p.CreateLogger(It.IsAny<string>()), Times.Never);
+    }
+
+    [TestMethod]
+    public async Task BuildAsync_WithDisabledExtensionProvider_DoesNotCallInitialize()
+    {
+        Mock<IExtensionInitializableLoggerProvider> mockProvider = new();
+        mockProvider.Setup(p => p.IsEnabledAsync()).ReturnsAsync(false);
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply test/UnitTests/Microsoft.Testing.Platform.UnitTests/Logging/LoggingManagerTests.cs:82]: @copilot address review comment

File: test/UnitTests/Microsoft.Testing.Platform.UnitTests/Helpers/ExtensionValidationHelperTests.cs [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    }
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply test/UnitTests/Microsoft.Testing.Platform.UnitTests/Helpers/ExtensionValidationHelperTests.cs:79]: @copilot address review comment

**Comment by Evangelink** [comment]
@copilot resolve the merge conflicts in this pull request

**Comment by Evangelink** [comment]
---
> [!NOTE]
> The pull request was not created — a fallback review issue was created instead due to protected file changes: [#8140](https://github.com/microsoft/testfx/issues/8140)

🤖 *Test Improver here.* Status update: the work in this PR has been superseded. The LoggingManager tests landed in main via #8124 and #8130, and the ExtensionValidationHelper tests landed via #8128.

This PR is a duplicate of merged work — it can be safely closed.

> Generated by [Daily Test Improver](https://github.com/microsoft/testfx/actions/runs/25704798051/agentic_workflow) · ● 9.5M · [◷](https://github.com/search?q=repo%3Amicrosoft%2Ftestfx+%22gh-aw-workflow-call-id%3A+microsoft%2Ftestfx%2Fdaily-test-improver%22&type=issues)
